File: tools/cmux-home/typescript/src/state.ts
import {
  adapterOrder,
  adapters,
  buildResumeCommand,
  compareAdapters,
  normalizeAdapterId,
  shellQuote,
  type AdapterId,
} from "./adapters";

export const statusOrder = ["awaiting", "working", "completed"] as const;

export type SessionStatus = (typeof statusOrder)[number];

export interface HomeSession {
  id: string;
  adapter: AdapterId;
  sessionId?: string;
  status: SessionStatus;
  title: string;
  cwd?: string;
  branch?: string;
  updatedAt?: string;
  preview?: string;
  details?: string;
  task?: string;
  resumeCommand?: string;
}

export interface HomeState {
  generatedAt?: string;
  sessions: HomeSession[];
}

export interface GroupedSessions {
  status: SessionStatus;
  sessions: HomeSession[];
}

interface SessionContext {
  cwd?: string;
  branch?: string;
  task?: string;
}

export function parseHomeState(raw: unknown): HomeState {
  validateSchemaContract(raw);
  const root = recordValue(raw);
  const generatedAt = stringValue(root?.generatedAt ?? root?.generated_at ?? root?.updatedAt ?? root?.updated_at);
  const rawSessions = collectRawSessions(raw);
  const sessions = rawSessions
    .map(({ value, context }, index) => parseSession(value, index, context))
    .filter((session): session is HomeSession => Boolean(session))
    .sort(compareSessions);
  return { generatedAt, sessions };
}

export function createFallbackState(): HomeState {
  return parseHomeState({
    generatedAt: "2026-01-01T00:00:00.000Z",
    sessions: [
      {
        id: "claude-plan",
        adapter: "claude",
        sessionId: "claude-session-123",
        status: "awaiting",
        title: "Review plan for terminal notifications",
        cwd: "/Users/example/cmux",
        branch: "feat-agent-home",
        preview: "Claude is waiting for plan approval.",
        details: "Prototype row that mirrors a cmux Feed plan/permission stop.",
      },
      {
        id: "codex-fix",
        adapter: "codex",
        sessionId: "codex-session-456",
        status: "working",
        title: "Implement grouped home state parser",
        cwd: "/Users/example/cmux",
        branch: "feat-agent-home",
        preview: "Codex is editing the TypeScript prototype.",
      },
      {
        id: "opencode-completed",
        adapter: "opencode",
        sessionId: "opencode-session-789",
        status: "completed",
        title: "Investigate OpenCode resume",
        cwd: "/Users/example/cmux",
        preview: "OpenCode has a resumable session.",
      },
      {
        id: "pi-completed",
        adapter: "pi",
        sessionId: "pi-session-101",
        status: "completed",
        title: "Sketch Vault registry bridge",
        cwd: "/Users/example/cmux",
        preview: "Pi finished its first pass.",
      },
    ],
  });
}

export function groupSessionsByStatus(sessions: HomeSession[]): GroupedSessions[] {
  return statusOrder.map((status) => ({
    status,
    sessions: sessions.filter((session) => session.status === status).sort(compareSessions),
  }));
}

export function adapterCounts(sessions: HomeSession[]): Record<AdapterId, number> {
  const counts = Object.fromEntries(adapterOrder.map((adapter) => [adapter, 0])) as Record<AdapterId, number>;
  for (const session of sessions) {
    counts[session.adapter] += 1;
  }
  return counts;
}

export function statusCounts(sessions: HomeSession[]): Record<SessionStatus, number> {
  const counts = Object.fromEntries(statusOrder.map((status) => [status, 0])) as Record<SessionStatus, number>;
  for (const session of sessions) {
    counts[session.status] += 1;
  }
  return counts;
}

export function compareSessions(left: HomeSession, right: HomeSession): number {
  const statusDelta = statusOrder.indexOf(left.status) - statusOrder.indexOf(right.status);
  if (statusDelta !== 0) {
    return statusDelta;
  }
  const adapterDelta = compareAdapters(left.adapter, right.adapter);
  if (adapterDelta !== 0) {
    return adapterDelta;
  }
  const updatedDelta = timestamp(right.updatedAt) - timestamp(left.updatedAt);
  if (updatedDelta !== 0) {
    return updatedDelta;
  }
  const titleDelta = left.title.localeCompare(right.title);
  if (titleDelta !== 0) {
    return titleDelta;
  }
  return left.id.localeCompare(right.id);
}

function parseSession(raw: unknown, index: number, context: SessionContext): HomeSession | undefined {
  const record = recordValue(raw);
  if (!record) {
    return undefined;
  }

  const agentRecord = recordValue(record.agent);
  const terminalRecord = recordValue(record.terminal);
  const terminalAgentRecord = recordValue(terminalRecord?.agent);
  const merged = {
    ...terminalAgentRecord,
    ...agentRecord,
    ...record,
  };

  const adapter = normalizeAdapterId(
    merged.adapter ?? merged.agent ?? merged.source ?? merged.kind ?? merged.agentKind ?? merged.agent_kind,
  );
  if (!adapter) {
    return undefined;
  }

  const sessionId = stringValue(
    merged.agentSessionId
      ?? merged.sessionId
      ?? merged.agent_session_id
      ?? merged.session_id
      ?? merged.nativeSessionId
      ?? merged.native_session_id
      ?? merged.chatId,
  );
  const id = stringValue(merged.id ?? merged.panelId ?? merged.surfaceId ?? merged.workspaceId)
    ?? `${adapter}-${sessionId ?? index + 1}`;
  const cwd = stringValue(
    merged.cwd
      ?? merged.workingDirectory
      ?? merged.working_directory
      ?? merged.workspacePath
      ?? recordValue(merged.workspace)?.cwd
      ?? context.cwd,
  );
  const title = stringValue(merged.title ?? merged.name ?? merged.summary ?? merged.task ?? merged.prompt)
    ?? `${adapters[adapter].displayName} session`;
  const workspaceGit = recordValue(recordValue(merged.workspace)?.git);
  const activity = recordValue(merged.activity);
  const attention = recordValue(merged.attention);
  const resume = recordValue(merged.resume);
  const status = normalizeStatus(merged.status ?? merged.state ?? merged.phase ?? activity?.phase);
  const resumeCommand = stringValue(merged.resumeCommand ?? merged.resume_command)
    ?? commandArrayValue(resume?.command)
    ?? buildResumeCommand({
      adapter,
      sessionId,
      cwd,
      model: stringValue(merged.model ?? merged.providerModel ?? merged.provider_model),
      permissionMode: stringValue(merged.permissionMode ?? merged.permission_mode),
      approvalPolicy: stringValue(merged.approvalPolicy ?? merged.approval_policy),
      sandboxMode: stringValue(merged.sandboxMode ?? merged.sandbox_mode),
      effort: stringValue(merged.effort ?? merged.reasoningEffort ?? merged.reasoning_effort),
      agentName: stringValue(merged.agentName ?? merged.agent_name),
      thinking: stringValue(merged.thinking),
    });

  return {
    id,
    adapter,
    sessionId,
    status,
    title,
    cwd,
    branch: stringValue(merged.gitBranch ?? merged.git_branch ?? merged.branch ?? workspaceGit?.branch ?? context.branch),
    updatedAt: dateString(merged.updatedAt ?? merged.updated_at ?? merged.modified ?? merged.modifiedAt),
    preview: stringValue(
      merged.preview
        ?? merged.lastMessage
        ?? merged.last_message
        ?? activity?.lastMessage
        ?? attention?.promptSummary
        ?? merged.message
        ?? merged.notification,
    ),
    details: stringValue(merged.details ?? merged.detail ?? merged.summary ?? merged.description ?? merged.transcriptSummary),
    task: stringValue(merged.task ?? merged.currentTask ?? merged.current_task ?? context.task),
    resumeCommand,
  };
}

function collectRawSessions(raw: unknown): Array<{ value: unknown; context: SessionContext }> {
  if (Array.isArray(raw)) {
    return raw.map((value) => ({ value, context: {} }));
  }

  const root = recordValue(raw);
  if (!root) {
    return [];
  }

  const direct = arrayValue(root.sessions)
    ?? arrayValue(recordValue(root.home)?.sessions)
    ?? arrayValue(recordValue(root.state)?.sessions)
    ?? arrayValue(recordValue(root.cmuxHome)?.sessions);
  if (direct) {
    return direct.map((value) => ({ value, context: {} }));
  }

  const workspaces = arrayValue(root.workspaces ?? root.tabs ?? root.panes);
  if (!workspaces) {
    return [];
  }

  const collected: Array<{ value: unknown; context: SessionContext }> = [];
  for (const workspace of workspaces) {
    const workspaceRecord = recordValue(workspace);
    const context: SessionContext = {
      cwd: stringValue(workspaceRecord?.cwd ?? workspaceRecord?.workingDirectory ?? workspaceRecord?.path),
      branch: stringValue(workspaceRecord?.gitBranch ?? workspaceRecord?.git_branch ?? workspaceRecord?.branch),
      task: stringValue(workspaceRecord?.task ?? workspaceRecord?.currentTask),
    };
    const nested = [
      ...(arrayValue(workspaceRecord?.sessions) ?? []),
      ...(arrayValue(workspaceRecord?.panels) ?? []),
      ...(arrayValue(workspaceRecord?.surfaces) ?? []),
      ...(arrayValue(workspaceRecord?.tabs) ?? []),
    ];
    for (const value of nested) {
      collected.push({ value, context });
    }
  }
  return collected;
}

function normalizeStatus(value: unknown): SessionStatus {
  if (typeof value !== "string") {
    return "completed";
  }
  const normalized = value.trim().toLowerCase().replace(/[\s-]+/g, "_");
  switch (normalized) {
    case "waiting":
    case "awaiting":
    case "needs_input":
    case "needs_user_input":
    case "awaitinguser":
    case "awaiting_user":
    case "blocked":
    case "pending":
      return "awaiting";
    case "running":
    case "active":
    case "working":
    case "executing":
    case "busy":
    case "in_progress":
      return "working";
    case "idle":
    case "ready":
    case "paused":
    case "stopped":
      return "completed";
    case "done":
    case "complete":
    case "completed":
    case "success":
    case "succeeded":
      return "completed";
    case "failed":
    case "error":
    case "crashed":
    case "cancelled":
    case "canceled":
      return "awaiting";
    default:
      return "completed";
  }
}

function validateSchemaContract(raw: unknown): void {
  const root = recordValue(raw);
  const version = root?.schemaVersion ?? root?.schema_version;
  if (version === undefined) {
    return;
  }
  if (version !== 1) {
    throw new Error("unsupported cmux home schemaVersion, expected 1");
  }
  const sessions = arrayValue(root?.sessions);
  for (const [index, session] of sessions?.entries() ?? []) {
    const status = recordValue(session)?.status;
    if (status !== "awaiting" && status !== "working" && status !== "completed") {
      throw new Error(`sessions[${index}].status must be awaiting, working, or completed`);
    }
  }
}

function recordValue(value: unknown): Record<string, unknown> | undefined {
  return value && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : undefined;
}

function arrayValue(value: unknown): unknown[] | undefined {
  return Array.isArray(value) ? value : undefined;
}

function commandArrayValue(value: unknown): string | undefined {
  const parts = arrayValue(value);
  const commandParts = parts?.map((part) => typeof part === "string" ? part.trim() : "");
  if (!commandParts || commandParts.some((part) => !part)) {
    return undefined;
  }
  return commandParts.map(shellQuote).join(" ");
}

function stringValue(value: unknown): string | undefined {
  if (typeof value !== "string") {
    return undefined;
  }
  const trimmed = value.trim();
  return trimmed || undefined;
}

function dateString(value: unknown): string | undefined {
  if (typeof value === "number" && Number.isFinite(value)) {
    return new Date(value > 10_000_000_000 ? value : value * 1_000).toISOString();
  }
  if (typeof value !== "string" || !value.trim()) {
    return undefined;
  }
  const numeric = Number(value);
  if (Number.isFinite(numeric)) {
    return dateString(numeric);
  }
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? undefined : date.toISOString();
}

function timestamp(value: string | undefined): number {
  if (!value) {
    return 0;
  }
  const parsed = Date.parse(value);
  return Number.isNaN(parsed) ? 0 : parsed;
}

File: tools/cmux-home/rust/src/state.rs
use crate::adapters::{adapter_sort_key, normalize_adapter_id, ADAPTERS};
use serde_json::Value;
use std::collections::{BTreeMap, HashMap};
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HomeState {
    pub sessions: Vec<Session>,
    pub task_prompt: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Session {
    pub id: String,
    pub adapter: &'static str,
    pub session_id: Option<String>,
    pub title: String,
    pub cwd: Option<String>,
    pub branch: Option<String>,
    pub status: SessionStatus,
    pub preview: Option<String>,
    pub details: Option<String>,
    pub updated_at: Option<String>,
    pub resume_command: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum SessionStatus {
    Awaiting,
    Working,
    Completed,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SessionGroup {
    pub status: SessionStatus,
    pub sessions: Vec<Session>,
}

impl HomeState {
    pub fn sorted_sessions(&self) -> Vec<Session> {
        let mut sessions = self.sessions.clone();
        sessions.sort_by(|left, right| {
            left.status
                .sort_key()
                .cmp(&right.status.sort_key())
                .then(adapter_sort_key(left.adapter).cmp(&adapter_sort_key(right.adapter)))
                .then(
                    left.title
                        .to_ascii_lowercase()
                        .cmp(&right.title.to_ascii_lowercase()),
                )
                .then(left.id.cmp(&right.id))
        });
        sessions
    }
}

impl Session {
    pub fn resume_session_id(&self) -> &str {
        self.session_id.as_deref().unwrap_or(&self.id)
    }
}

impl SessionStatus {
    pub const ORDERED: [SessionStatus; 3] = [
        SessionStatus::Awaiting,
        SessionStatus::Working,
        SessionStatus::Completed,
    ];

    pub fn label(self) -> &'static str {
        match self {
            SessionStatus::Awaiting => "awaiting",
            SessionStatus::Working => "working",
            SessionStatus::Completed => "completed",
        }
    }

    fn sort_key(self) -> usize {
        Self::ORDERED
            .iter()
            .position(|candidate| *candidate == self)
            .unwrap_or(usize::MAX)
    }
}

pub fn load_state(data_path: Option<&Path>) -> Result<HomeState, String> {
    if let Some(path) = data_path {
        let contents = fs::read_to_string(path)
            .map_err(|error| format!("failed to read {}: {error}", path.display()))?;
        return parse_state(&contents);
    }

    for path in default_state_candidates() {
        if path.is_file() {
            let contents = fs::read_to_string(&path)
                .map_err(|error| format!("failed to read {}: {error}", path.display()))?;
            return parse_state(&contents);
        }
    }

    Ok(fallback_state())
}

pub fn parse_state(contents: &str) -> Result<HomeState, String> {
    let value: Value = serde_json::from_str(contents).map_err(|error| error.to_string())?;
    validate_schema_contract(&value)?;
    let task_prompt = extract_task_prompt(&value)
        .unwrap_or_else(|| "Describe a task for the selected agent...".to_string());
    let mut session_values = Vec::new();
    collect_session_values(&value, &mut session_values);

    let mut sessions = Vec::new();
    for session_value in session_values {
        if let Some(session) = parse_session(session_value) {
            sessions.push(session);
        }
    }

    Ok(HomeState {
        sessions,
        task_prompt,
    })
}

pub fn fallback_state() -> HomeState {
    HomeState {
        task_prompt: "Ask an agent to summarize current branch risk...".to_string(),
        sessions: vec![
            Session {
                id: "claude-demo".to_string(),
                adapter: "claude",
                session_id: Some("claude-demo".to_string()),
                title: "Review sidebar notification polish".to_string(),
                cwd: Some("~/fun/cmux".to_string()),
                branch: Some("feat/home-prototype".to_string()),
                status: SessionStatus::Awaiting,
                preview: Some("Needs approval for edit plan".to_string()),
                details: Some("Claude Code is paused on an approval-style event.".to_string()),
                updated_at: Some("2026-05-12T09:30:00Z".to_string()),
                resume_command: None,
            },
            Session {
                id: "codex-demo".to_string(),
                adapter: "codex",
                session_id: Some("codex-demo".to_string()),
                title: "Implement Ratatui smoke mode".to_string(),
                cwd: Some("~/fun/cmux".to_string()),
                branch: Some("feat/cmux-home".to_string()),
                status: SessionStatus::Working,
                preview: Some("Running parser tests".to_string()),
                details: Some("Codex session is actively building the Rust prototype.".to_string()),
                updated_at: Some("2026-05-12T09:32:00Z".to_string()),
                resume_command: None,
            },
            Session {
                id: "opencode-demo".to_string(),
                adapter: "opencode",
                session_id: Some("opencode-demo".to_string()),
                title: "Sketch OpenCode adapter gaps".to_string(),
                cwd: Some("~/fun/cmux".to_string()),
                branch: None,
                status: SessionStatus::Completed,
                preview: Some("Ready to resume".to_string()),
                details: Some("OpenCode state came from fallback demo data.".to_string()),
                updated_at: Some("2026-05-12T09:20:00Z".to_string()),
                resume_command: None,
            },
            Session {
                id: "pi-demo".to_string(),
                adapter: "pi",
                session_id: Some("pi-demo".to_string()),
                title: "Validate Pi registry assumptions".to_string(),
                cwd: Some("~/fun/cmux".to_string()),
                branch: None,
                status: SessionStatus::Completed,
                preview: Some("Prototype note captured".to_string()),
                details: Some("Pi is shown as a read-only restorable agent.".to_string()),
                updated_at: Some("2026-05-12T09:10:00Z".to_string()),
                resume_command: None,
            },
        ],
    }
}

pub fn group_sessions_by_status(state: &HomeState) -> Vec<SessionGroup> {
    let mut grouped: BTreeMap<SessionStatus, Vec<Session>> = BTreeMap::new();
    for session in state.sorted_sessions() {
        grouped.entry(session.status).or_default().push(session);
    }

    SessionStatus::ORDERED
        .into_iter()
        .filter_map(|status| {
            let sessions = grouped.remove(&status)?;
            if sessions.is_empty() {
                None
            } else {
                Some(SessionGroup { status, sessions })
            }
        })
        .collect()
}

pub fn adapter_counts(state: &HomeState) -> Vec<(&'static str, usize)> {
    let counts = state.sessions.iter().fold(
        HashMap::<&'static str, usize>::new(),
        |mut counts, session| {
            *counts.entry(session.adapter).or_default() += 1;
            counts
        },
    );

    ADAPTERS
        .iter()
        .map(|adapter| {
            (
                adapter.id,
                counts.get(adapter.id).copied().unwrap_or_default(),
            )
        })
        .collect()
}

pub fn status_counts(state: &HomeState) -> Vec<(SessionStatus, usize)> {
    let counts = state.sessions.iter().fold(
        HashMap::<SessionStatus, usize>::new(),
        |mut counts, session| {
            *counts.entry(session.status).or_default() += 1;
            counts
        },
    );

    SessionStatus::ORDERED
        .into_iter()
        .map(|status| (status, counts.get(&status).copied().unwrap_or_default()))
        .collect()
}

fn default_state_candidates() -> Vec<PathBuf> {
    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    vec![
        manifest_dir.join("../state/example.json"),
        manifest_dir.join("../state/cmux-home.json"),
        manifest_dir.join("../example-state.json"),
        manifest_dir.join("example-state.json"),
        PathBuf::from("tools/cmux-home/state/example.json"),
        PathBuf::from("tools/cmux-home/state/cmux-home.json"),
    ]
}

fn validate_schema_contract(value: &Value) -> Result<(), String> {
    let Some(root) = value.as_object() else {
        return Ok(());
    };
    let Some(version) = root
        .get("schemaVersion")
        .or_else(|| root.get("schema_version"))
    else {
        return Ok(());
    };
    if version.as_u64() != Some(1) {
        return Err("unsupported cmux home schemaVersion, expected 1".to_string());
    }
    if let Some(sessions) = root.get("sessions").and_then(Value::as_array) {
        for (index, session) in sessions.iter().enumerate() {
            let status = session
                .as_object()
                .and_then(|session| session.get("status"))
                .and_then(Value::as_str)
                .ok_or_else(|| format!("sessions[{index}].status is required"))?;
            if !matches!(status, "awaiting" | "working" | "completed") {
                return Err(format!(
                    "sessions[{index}].status must be awaiting, working, or completed"
                ));
            }
        }
    }
    Ok(())
}

fn collect_session_values<'a>(value: &'a Value, out: &mut Vec<&'a Value>) {
    match value {
        Value::Array(items) => {
            for item in items {
                if looks_like_session(item) {
                    out.push(item);
                } else {
                    collect_session_values(item, out);
                }
            }
        }
        Value::Object(map) => {
            if looks_like_session(value) {
                out.push(value);
                return;
            }

            for key in ["sessions", "workstreams", "items", "agents", "state"] {
                if let Some(child) = map.get(key) {
                    collect_session_values(child, out);
                }
            }
        }
        _ => {}
    }
}

fn looks_like_session(value: &Value) -> bool {
    let Some(map) = value.as_object() else {
        return false;
    };
    let has_adapter = first_string(
        map,
        &["adapter", "agent", "source", "_source", "provider", "kind"],
    )
    .and_then(normalize_adapter_id)
    .is_some();
    let has_id = first_string(
        map,
        &[
            "agentSessionId",
            "agent_session_id",
            "sessionId",
            "session_id",
            "threadId",
            "thread_id",
            "workstreamId",
            "workstream_id",
            "id",
        ],
    )
    .is_some();

    has_adapter && has_id
}

fn parse_session(value: &Value) -> Option<Session> {
    let map = value.as_object()?;
    let workspace = object_field(map, "workspace");
    let workspace_git = workspace.and_then(|workspace| object_field(workspace, "git"));
    let resume = object_field(map, "resume");
    let activity = object_field(map, "activity");
    let attention = object_field(map, "attention");
    let adapter = first_string(
        map,
        &["adapter", "agent", "source", "_source", "provider", "kind"],
    )
    .and_then(normalize_adapter_id)?;
    let id = first_string(
        map,
        &[
            "id",
            "agentSessionId",
            "agent_session_id",
            "sessionId",
            "session_id",
            "threadId",
            "thread_id",
            "workstreamId",
            "workstream_id",
        ],
    )?
    .to_string();
    let session_id = first_string(
        map,
        &[
            "agentSessionId",
            "agent_session_id",
            "sessionId",
            "session_id",
            "threadId",
            "thread_id",
            "workstreamId",
            "workstream_id",
        ],
    )
    .map(ToString::to_string)
    .or_else(|| Some(id.clone()));
    let title = first_string(map, &["title", "name", "summary", "prompt"])
        .map(clean_line)
        .filter(|title| !title.is_empty())
        .unwrap_or_else(|| format!("{adapter} {}", session_id.as_deref().unwrap_or(&id)));

    Some(Session {
        id,
        adapter,
        session_id,
        title,
        cwd: first_string(
            map,
            &[
                "cwd",
                "workingDirectory",
                "working_directory",
                "repo",
                "workspace",
            ],
        )
        .or_else(|| workspace.and_then(|workspace| first_string(workspace, &["cwd"])))
        .map(ToString::to_string),
        branch: first_string(map, &["branch", "gitBranch", "git_branch"])
            .or_else(|| {
                workspace_git
                    .and_then(|git| first_string(git, &["branch", "gitBranch", "git_branch"]))
            })
            .map(ToString::to_string),
        status: parse_status(
            map.get("status")
                .or_else(|| map.get("state"))
                .or_else(|| activity.and_then(|activity| activity.get("phase"))),
        ),
        preview: first_string(
            map,
            &[
                "preview",
                "lastMessage",
                "last_message",
                "message",
                "body",
                "text",
                "prompt",
            ],
        )
        .or_else(|| {
            activity.and_then(|activity| first_string(activity, &["lastMessage", "last_message"]))
        })
        .or_else(|| {
            attention
                .and_then(|attention| first_string(attention, &["promptSummary", "prompt_summary"]))
        })
        .map(clean_line)
        .filter(|value| !value.is_empty()),
        details: first_string(
            map,
            &[
                "details",
                "detail",
                "transcript",
                "lastAssistantMessage",
                "last_assistant_message",
                "description",
            ],
        )
        .map(clean_line)
        .filter(|value| !value.is_empty()),
        updated_at: first_string(
            map,
            &[
                "updatedAt",
                "updated_at",
                "modified",
                "lastUpdated",
                "last_updated",
            ],
        )
        .map(ToString::to_string),
        resume_command: resume
            .and_then(|resume| resume.get("command"))
            .and_then(command_array_value),
    })
}

fn parse_status(value: Option<&Value>) -> SessionStatus {
    match value {
        Some(Value::String(value)) => parse_status_string(value),
        Some(Value::Object(map)) => {
            if map.contains_key("pending") {
                SessionStatus::Awaiting
            } else if map.contains_key("resolved") || map.contains_key("expired") {
                SessionStatus::Completed
            } else if map.contains_key("telemetry") {
                SessionStatus::Working
            } else {
                map.keys()
                    .next()
                    .map(|key| parse_status_string(key))
                    .unwrap_or(SessionStatus::Completed)
            }
        }
        Some(Value::Bool(true)) => SessionStatus::Working,
        _ => SessionStatus::Completed,
    }
}

fn parse_status_string(value: &str) -> SessionStatus {
    match value.trim().to_ascii_lowercase().replace('_', "-").as_str() {
        "awaiting" | "awaiting-user" | "awaitinguser" | "awaiting user" | "waiting" | "pending"
        | "needs-input" | "needs input" | "blocked" | "paused" => SessionStatus::Awaiting,
        "running" | "active" | "busy" | "working" | "in-progress" | "telemetry" => {
            SessionStatus::Working
        }
        "idle" | "stopped" | "ready" => SessionStatus::Completed,
        "done" | "complete" | "completed" | "finished" | "resolved" | "expired" => {
            SessionStatus::Completed
        }
        "failed" | "failure" | "error" | "crashed" => SessionStatus::Awaiting,
        _ => SessionStatus::Completed,
    }
}

fn object_field<'a>(
    map: &'a serde_json::Map<String, Value>,
    key: &str,
) -> Option<&'a serde_json::Map<String, Value>> {
    map.get(key).and_then(Value::as_object)
}

fn command_array_value(value: &Value) -> Option<String> {
    let values = value.as_array()?;
    let mut parts = Vec::with_capacity(values.len());
    for value in values {
        let part = value_as_string(value)?.trim();
        if part.is_empty() {
            return None;
        }
        parts.push(shell_word(part));
    }
    Some(parts.join(" "))
}

fn shell_word(value: &str) -> String {
    if value.chars().all(|ch| {
        ch.is_ascii_alphanumeric() || matches!(ch, '_' | '.' | '/' | ':' | '=' | '+' | '-')
    }) {
        return value.to_string();
    }
    format!("'{}'", value.replace('\'', "'\\''"))
}

fn first_string<'a>(map: &'a serde_json::Map<String, Value>, keys: &[&str]) -> Option<&'a str> {
    for key in keys {
        if let Some(value) = map.get(*key).and_then(value_as_string) {
            return Some(value);
        }
    }
    None
}

fn value_as_string(value: &Value) -> Option<&str> {
    match value {
        Value::String(value) => Some(value),
        _ => None,
    }
}

fn clean_line(value: &str) -> String {
    value.split_whitespace().collect::<Vec<_>>().join(" ")
}

fn extract_task_prompt(value: &Value) -> Option<String> {
    match value {
        Value::Object(map) => first_string(
            map,
            &[
                "taskPrompt",
                "task_prompt",
                "promptPlaceholder",
                "prompt_placeholder",
            ],
        )
        .map(ToString::to_string)
        .or_else(|| map.get("state").and_then(extract_task_prompt)),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_root_sessions_shape() {
        let state = parse_state(
            r#"{
              "taskPrompt": "Ask an agent...",
              "sessions": [
                {
                  "id": "item-1",
                  "session_id": "claude-session",
                  "source": "claude",
                  "status": "pending",
                  "title": "Fix tabs",
                  "cwd": "/tmp/cmux",
                  "branch": "feat/tabs",
                  "preview": "Needs approval"
                }
              ]
            }"#,
        )
        .unwrap();

        assert_eq!(state.task_prompt, "Ask an agent...");
        assert_eq!(state.sessions.len(), 1);
        assert_eq!(state.sessions[0].adapter, "claude");
        assert_eq!(state.sessions[0].status, SessionStatus::Awaiting);
        assert_eq!(state.sessions[0].resume_session_id(), "claude-session");
        assert_eq!(state.sessions[0].preview.as_deref(), Some("Needs approval"));
    }

    #[test]
    fn parses_nested_state_shape() {
        let state = parse_state(
            r#"{
              "state": {
                "sessions": [
                  {"id": "codex-1", "agent": "codex", "status": "running", "title": "Build it"}
                ]
              }
            }"#,
        )
        .unwrap();

        assert_eq!(state.sessions.len(), 1);
        assert_eq!(state.sessions[0].adapter, "codex");
        assert_eq!(state.sessions[0].status, SessionStatus::Working);
    }

    #[test]
    fn groups_sessions_by_status_in_home_order() {
        let state = HomeState {
            task_prompt: "Task".to_string(),
            sessions: vec![
                Session {
                    id: "completed".to_string(),
                    adapter: "pi",
                    session_id: None,
                    title: "Completed".to_string(),
                    cwd: None,
                    branch: None,
                    status: SessionStatus::Completed,
                    preview: None,
                    details: None,
                    updated_at: None,
                    resume_command: None,
                },
                Session {
                    id: "awaiting".to_string(),
                    adapter: "claude",
                    session_id: None,
                    title: "Awaiting".to_string(),
                    cwd: None,
                    branch: None,
                    status: SessionStatus::Awaiting,
                    preview: None,
                    details: None,
                    updated_at: None,
                    resume_command: None,
                },
            ],
        };

        let groups = group_sessions_by_status(&state);

        assert_eq!(
            groups.iter().map(|group| group.status).collect::<Vec<_>>(),
            vec![SessionStatus::Awaiting, SessionStatus::Completed]
        );
    }

    #[test]
    fn counts_all_known_adapters() {
        let state = parse_state(
            r#"[
              {"id": "c1", "adapter": "claude", "status": "idle"},
              {"id": "c2", "adapter": "codex", "status": "idle"}
            ]"#,
        )
        .unwrap();

        assert_eq!(
            adapter_counts(&state),
            vec![("claude", 1), ("codex", 1), ("opencode", 0), ("pi", 0)]
        );
    }

    #[test]
    fn rejects_unsupported_schema_version() {
        let error = parse_state(r#"{"schemaVersion":2,"sessions":[]}"#).unwrap_err();

        assert!(error.contains("schemaVersion"));
    }

    #[test]
    fn rejects_schema_status_outside_contract() {
        let error = parse_state(
            r#"{
              "schemaVersion": 1,
              "sessions": [
                {"id": "bad", "agent": "codex", "agentSessionId": "bad", "status": "failed"}
              ]
            }"#,
        )
        .unwrap_err();

        assert!(error.contains("status"));
    }

    #[test]
    fn parses_canonical_schema_fields() {
        let state = parse_state(
            r#"{
              "schemaVersion": 1,
              "sessions": [
                {
                  "id": "codex:canonical",
                  "agent": "codex",
                  "agentSessionId": "canonical",
                  "title": "Canonical",
                  "status": "awaiting",
                  "updatedAt": "2026-05-12T16:00:00Z",
                  "workspace": {
                    "id": "workspace-id",
                    "cwd": "/tmp/cmux",
                    "git": {"branch": "feat/canonical"}
                  },
                  "activity": {"phase": "awaitingUser", "lastMessage": "Permission requested"},
                  "attention": {"promptSummary": "Run tests"},
                  "resume": {"command": ["codex", "resume", "canonical"]}
                }
              ]
            }"#,
        )
        .unwrap();

        let session = &state.sessions[0];
        assert_eq!(session.resume_session_id(), "canonical");
        assert_eq!(session.cwd.as_deref(), Some("/tmp/cmux"));
        assert_eq!(session.branch.as_deref(), Some("feat/canonical"));
        assert_eq!(session.preview.as_deref(), Some("Permission requested"));
        assert_eq!(
            session.resume_command.as_deref(),
            Some("codex resume canonical")
        );
    }
}
